src/main/scala/rise/core/primitives.scala

```diff
@@ -10,296 +10,303 @@ import rise.macros.Primitive.primitive
       (s ->: s) ->: // function to write a value
-        ((n + sz) `.` s) ->: ((1 + n) `.` sz `.` s))))
+        ((n + sz) `.` s) ->: ((1 + n) `.` sz `.` s) })}
   }
 
   @primitive object snd extends Primitive with Builder {
-    implDT(s => implDT(t => (s x t) ->: t))
+    impl{ s: DataType => impl{ t: DataType => (s x t) ->: t }}
   }
 
   @primitive object split extends Primitive with Builder {
-    forallNat(n => implNat(m => implDT(t =>
-      ((m * n)`.`t) ->: (m`.`n`.`t))))
+    expl( (n: Nat) => impl{ m: Nat => impl{ t: DataType =>
+      ((m * n)`.`t) ->: (m`.`n`.`t) }})
   }
 
   @primitive object take extends Primitive with Builder {
-    forallNat(n => implNat(m => implDT(t =>
-      ((n + m)`.`t) ->: (n`.`t))))
+    expl( (n: Nat) => impl{ m: Nat => impl{ t: DataType =>
+      ((n + m)`.`t) ->: (n`.`t) }})
   }
 
   @primitive object transpose extends Primitive with Builder {
-    implNat(n => implNat(m => implDT(dt =>
-      (n`.`m`.`dt) ->: (m`.`n`.`dt))))
+    impl{ n: Nat => impl{ m: Nat => impl{ dt: DataType =>
+      (n`.`m`.`dt) ->: (m`.`n`.`dt) }}}
   }
 
   // if-then-else
   @primitive object select extends Primitive with Builder {
-    implDT(t => bool ->: t ->: t ->: t)
+    impl{ t: DataType => bool ->: t ->: t ->: t }
   }
 
   @primitive object unzip extends Primitive with Builder {
-    implNat(n => implDT(dt1 => implDT(dt2 =>
-      (n`.`(dt1 x dt2)) ->: ((n`.`dt1) x (n`.`dt2)))))
+    impl{ n: Nat => impl{ dt1: DataType => impl{ dt2: DataType =>
+      (n`.`(dt1 x dt2)) ->: ((n`.`dt1) x (n`.`dt2)) }}}
   }
 
   @primitive object zip extends Primitive with Builder {
-    implNat(n => implDT(a => implDT(b =>
-      (n`.`a) ->: (n`.`b) ->: (n`.`(a x b)))))
+    impl{ n: Nat => impl{ a: DataType => impl{ b: DataType =>
+      (n`.`a) ->: (n`.`b) ->: (n`.`(a x b)) }}}
   }
 
   @primitive object neg extends Primitive with Builder {
-    implBT(t => t ->: t)
+    impl{ t: DataType => t ->: t }
   }
 
   @primitive object not extends Primitive with Builder {
     bool ->: bool
   }
 
-  @primitive object add extends Primitive with Builder { implBT(t => t ->: t ->: t) }
-  @primitive object sub extends Primitive with Builder { implBT(t => t ->: t ->: t) }
-  @primitive object mul extends Primitive with Builder { implBT(t => t ->: t ->: t) }
-  @primitive object div extends Primitive with Builder { implBT(t => t ->: t ->: t) }
-  @primitive object mod extends Primitive with Builder { implBT(t => t ->: t ->: t) }
+  @primitive object add extends Primitive with Builder { impl{ t: DataType => t ->: t ->: t } }
+  @primitive object sub extends Primitive with Builder { impl{ t: DataType => t ->: t ->: t } }
+  @primitive object mul extends Primitive with Builder { impl{ t: DataType => t ->: t ->: t } }
+  @primitive object div extends Primitive with Builder { impl{ t: DataType => t ->: t ->: t } }
+  @primitive object mod extends Primitive with Builder { impl{ t: DataType => t ->: t ->: t } }
```

Comment:
These primitives should only work on "basic types" (scalars/vectors). https://github.com/rise-lang/rise/issues/11
Maybe we can leave a TODO in the code at least?